diff --git a/dom/canvas/ClientWebGLContext.cpp b/dom/canvas/ClientWebGLContext.cpp
--- a/dom/canvas/ClientWebGLContext.cpp
+++ b/dom/canvas/ClientWebGLContext.cpp
@@ -3695,7 +3695,9 @@
       if (extraction == CanvasUtils::ImageExtraction::Placeholder) {
         dom::GeneratePlaceholderCanvasData(destView->size_bytes(),
                                            destView->Elements());
-      } else if (extraction == CanvasUtils::ImageExtraction::Randomize) {
+      } else if (extraction == CanvasUtils::ImageExtraction::Randomize ||
+                 extraction ==
+                     CanvasUtils::ImageExtraction::EfficientRandomize) {
         // We have no idea what's in the buffer. So, we randomize it as if each
         // elemSize bytes is a single element.
         uint8_t elementsPerGroup = 1,
@@ -5417,7 +5419,8 @@
       } else {
         RecordCanvasUsage(CanvasExtractionAPI::ReadPixels,
                           CSSIntSize(width, height));
-        if (extraction == CanvasUtils::ImageExtraction::Randomize) {
+        if (extraction == CanvasUtils::ImageExtraction::Randomize ||
+            extraction == CanvasUtils::ImageExtraction::EfficientRandomize) {
           const auto pii = webgl::PackingInfoInfo::For(desc.pi);
           // DoReadPixels() requres pii to be Some().
           MOZ_ASSERT(pii.isSome());
